File: All missing numbers.py
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        #Time Complexity: O(n)
        #Space Complexity: O(1)
        #where, n is the size of nums
        
        for num in nums:
            idx = abs(num) - 1
            nums[idx] = - (abs(nums[idx]))
            
        return [i+1 for i in range(len(nums)) if nums[i] > 0]

File: AllMissingNumbers.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Approach: create a boolean array and mark the indexes for
// which we found the numbers.

// LC- 448. Find All Numbers Disappeared in an Array

import java.util.*;

class AllMissingNumbers {
  public List<Integer> findDisappearedNumbers(int[] nums) {
    List<Integer> result = new ArrayList<>();

    if (nums == null || nums.length == 0) {
      return result;
    }

    boolean[] array = new boolean[nums.length + 1];
    array[0] = true;

    for (int i = 0; i < nums.length; i++) {
      array[nums[i]] = true;
    }

    for (int i = 1; i < nums.length + 1; i++) {
      if (!array[i]) {
        result.add(i);
      }
    }

    return result;
  }
}